masonry/src/tree_arena.rs

```diff
@@ -138,92 +134,115 @@ impl<Item> TreeArena<Item> {
     ///
-    /// Returns a tuple of a mutable reference to the item, and a token to access
-    /// its children.
+    /// Returns a mutable reference to the item.
     ///
     /// ## Complexity
     ///
-    /// O(N) where N is the size of the tree. In future versions, will be O(depth)
-    /// or O(1).
+    /// O(Depth). In future implementations, this will be O(1).
     pub fn find_mut(&mut self, id: u64) -> Option<ArenaMut<'_, Item>> {
-        if !self.parents_map.contains_key(&id) {
-            return None;
-        }
-
-        let mut id_path = self.get_id_path(id);
-        let root_id = id_path.pop().unwrap();
-
-        let mut id_path = id_path.as_slice();
-        let mut node = self
-            .roots
-            .iter_mut()
-            .find(|root| root.id == root_id)
-            .unwrap();
-
-        while let Some((id, new_id_path)) = id_path.split_last() {
-            id_path = new_id_path;
-            node = node
-                .children
-                .iter_mut()
-                .find(|child| child.id == *id)
-                .unwrap();
-        }
-
-        Some(node.arena_mut(self.parents_map[&id], &mut self.parents_map))
+        self.root_token_mut().find_mut(id)
     }
 
     /// Construct the path of items from the given item to the root of the tree.
     ///
     /// The path is in order from the bottom to the top, starting at the given item and ending at
     /// the root.
+    ///
+    /// If the id is not in the tree, returns an empty vector.
     pub fn get_id_path(&self, id: u64) -> Vec<u64> {
-        let mut path = Vec::new();
+        let parents_map = ArenaMapRef {
+            parents_map: &self.parents_map,
+        };
+        parents_map.get_id_path(id, None)
+    }
+}
 
-        if !self.parents_map.contains_key(&id) {
-            return path;
+impl<Item> TreeNode<Item> {
+    fn arena_ref<'a>(
+        &'a self,
+        parent_id: Option<u64>,
+        parents_map: &'a HashMap<u64, Option<u64>>,
+    ) -> ArenaRef<'a, Item> {
+        ArenaRef {
+            parent_id,
+            item: &self.item,
+            children: ArenaRefChildren {
+                id: Some(self.id),
+                children: &self.children,
+                parents_map: ArenaMapRef { parents_map },
+            },
         }
+    }
 
-        let mut current_id = Some(id);
-        while let Some(id) = current_id {
-            path.push(id);
-            current_id = *self.parents_map.get(&id).unwrap();
+    fn arena_mut<'a>(
+        &'a mut self,
+        parent_id: Option<u64>,
+        parents_map: &'a mut HashMap<u64, Option<u64>>,
+    ) -> ArenaMut<'a, Item> {
+        ArenaMut {
+            parent_id,
+            item: &mut self.item,
+            children: ArenaMutChildren {
+                id: Some(self.id),
+                children: &mut self.children,
+                parents_map: ArenaMapMut { parents_map },
+            },
+        }
+    }
+}
+
+impl<'a, Item> ArenaRef<'a, Item> {
```

Comment:
I think the usual style is to put impls (especially trivial impls) near to the corresponding struct definition.

Reply:
I like having all the type declarations together. It's the style the rest of the crate follows.